lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,227 @@
+                    throw new \Aimeos\MShop\Service\Exception($response->getMessage());
+                }
+            } elseif ($response->isRedirect()) {
+                $this->setOrderData($order, ['Transaction' => $response->getTransactionReference()]);
+                return $this->getRedirectForm($response);
+            } else {
+                $this->saveOrder($order->setPaymentStatus(Status::PAY_REFUSED));
+                throw new \Aimeos\MShop\Service\Exception($response->getMessage());
+            }
+        } catch (\Exception $e) {
+            throw new \Aimeos\MShop\Service\Exception($e->getMessage());
+        }
+        // end of preparation section to call paypal plus */
+
+        return new \Aimeos\MShop\Common\Helper\Form\Standard('', '', [], true, $this->getPayPalPlusJs($approvalUrl, $countryid));
+    }
+
+    protected function getPayPalPlusJs($approvalUrl, $countryid)
+    {
+        $testmode = $this->getConfigValue('testmode') ? 'sandbox' : 'live';
+        return '
+		<script src="https://www.paypalobjects.com/webstatic/ppplus/ppplus.min.js" type="text/javascript">
+		</script>
+		<div id="ppplus">
+		</div>
+		<script type="application/javascript">		  
+		var approvalUrl = \'' . $approvalUrl . '\'; 
+		var ppp = PAYPAL.apps.PPP({
+			"approvalUrl": approvalUrl,
+			"placeholder": "ppplus",
+			"country":"' . $countryid . '" ,
+			"mode": "' . $testmode . '",
+			onContinue: function () { ppp.doCheckout(); } ,
+		});
+		</script>
+		';
+    }
+
+
+    public function process(\Aimeos\MShop\Order\Item\Iface $order, array $params = []): ?\Aimeos\MShop\Common\Helper\Form\Iface
+    {
+        return $this->getPaymentForm($order, $params);
+    }
+
+    // protected function getData(\Aimeos\MShop\Order\Item\Base\Iface $base, string $orderid, array $params): array
+    // {
+    //  return ['payerid' => "payerid-" .$this->getContext()->getUserId() ?: "payerid-" . $orderid] + parent::getData( $base , $orderid , $params);
+    // }
+
+
+    public function updateSync(
+        \Psr\Http\Message\ServerRequestInterface $request,
+        \Aimeos\MShop\Order\Item\Iface $order
+    ): \Aimeos\MShop\Order\Item\Iface {
+        try {
+            $provider = $this->getProvider();
+            $base = $this->getOrderBase($order->getBaseId());
+
+            $params = (array) $request->getAttributes() + (array) $request->getParsedBody() + (array) $request->getQueryParams();
+            $params = $this->getData($base, $order->getId(), $params);
+            $params['transactionReference'] = $this->getTransactionReference($base);
+
+            if ($this->getValue('authorize', false) && $provider->supportsCompleteAuthorize()) {
+                $response = $provider->completeAuthorize($params)->send();
+                $status = Status::PAY_AUTHORIZED;
+            } elseif ($provider->supportsCompletePurchase()) {
+                $params['PayerID'] = $request->getQueryParams()['PayerID'] ;
```

Comment:
I would push this to getData() nevertheless because as soon as your PR for `$response->getTransactionId()` to return the order ID and PAYMENT_NOT_APPROVED_FOR_EXECUTION leads to "cancelled", we can remove updateSync() completely.